pkg/controller/failurerecovery/pod_termination_controller.go

```diff
@@ -0,0 +1,168 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package failurerecovery
+
+import (
+	"context"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	configapi "sigs.k8s.io/kueue/apis/config/v1beta2"
+	utilnode "sigs.k8s.io/kueue/pkg/util/node"
+	utilpod "sigs.k8s.io/kueue/pkg/util/pod"
+)
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=update;patch
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+var (
+	realClock = clock.RealClock{}
+)
+
+type TerminatingPodReconciler struct {
+	client          client.Client
+	terminationCfgs []*terminatePodConfigWithSelector
+	clock           clock.Clock
+}
+
+type terminatePodConfigWithSelector struct {
+	selector       labels.Selector
+	terminationCfg configapi.TerminatePodConfig
+}
+
+type TerminatingPodReconcilerOptions struct {
+	clock clock.Clock
+}
+
+type TerminatingPodReconcilerOption func(*TerminatingPodReconcilerOptions)
+
+func WithClock(c clock.Clock) TerminatingPodReconcilerOption {
+	return func(o *TerminatingPodReconcilerOptions) {
+		o.clock = c
+	}
+}
+
+var defaultOptions = TerminatingPodReconcilerOptions{
+	clock: realClock,
+}
+
+func NewTerminatingPodReconciler(
+	client client.Client,
+	cfgs []configapi.TerminatePodConfig,
+	opts ...TerminatingPodReconcilerOption,
+) (*TerminatingPodReconciler, error) {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	terminationCfgs := make([]*terminatePodConfigWithSelector, len(cfgs))
+	for i, cfg := range cfgs {
+		selector, err := metav1.LabelSelectorAsSelector(cfg.PodLabelSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		terminationCfgs[i] = &terminatePodConfigWithSelector{
+			terminationCfg: cfg,
+			selector:       selector,
+		}
+	}
+
+	return &TerminatingPodReconciler{
+		client:          client,
+		terminationCfgs: terminationCfgs,
+		clock:           options.clock,
+	}, nil
+}
+
+func (r *TerminatingPodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	pod := &corev1.Pod{}
+	if err := r.client.Get(ctx, req.NamespacedName, pod); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	terminationCfg := firstMatchingTerminationConfig(r.terminationCfgs, pod)
+	// Pod does not match any of the configured pod label selector
+	if terminationCfg == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Pod was not marked for termination
+	if pod.DeletionTimestamp.IsZero() {
```

Comment:
In the KEP, you also planned to ignore Pods with `pod.DeletionGracePeriodSeconds == nil`.

Reply:
I think it doesn't really matter whether this is checked or not, since AFAIK K8s always set this to a non-nil value for pods that are terminating. Since it has to be dereferenced later anyway, I went for `ptr.Deref(pod.DeletionGracePeriodSeconds, 0)` just to be on the safe side. We could do both.

WDYT?